src/treemap/ops.rs

```diff
@@ -170,208 +145,320 @@ impl RoaringTreemap {
+                Some(other_rb) => {
+                    BitAndAssign::bitand_assign(self_rb, other_rb);
+                    if self_rb.is_empty() {
+                        keys_to_remove.push(*key);
+                    }
+                }
+                None => keys_to_remove.push(*key),
+            }
+        }
+
+        for key in keys_to_remove {
+            self.map.remove(&key);
+        }
     }
 }
 
 impl Sub<RoaringTreemap> for RoaringTreemap {
     type Output = RoaringTreemap;
 
+    /// This is equivalent to a `difference` between both maps.
     fn sub(mut self, rhs: RoaringTreemap) -> RoaringTreemap {
-        self.difference_with(&rhs);
+        SubAssign::sub_assign(&mut self, rhs);
         self
     }
 }
 
-impl<'a> Sub<&'a RoaringTreemap> for RoaringTreemap {
+impl Sub<&RoaringTreemap> for RoaringTreemap {
     type Output = RoaringTreemap;
 
-    fn sub(mut self, rhs: &'a RoaringTreemap) -> RoaringTreemap {
-        self.difference_with(rhs);
+    /// This is equivalent to a `difference` between both maps.
+    fn sub(mut self, rhs: &RoaringTreemap) -> RoaringTreemap {
+        SubAssign::sub_assign(&mut self, rhs);
         self
     }
 }
 
-impl<'a> Sub<RoaringTreemap> for &'a RoaringTreemap {
+impl Sub<RoaringTreemap> for &RoaringTreemap {
     type Output = RoaringTreemap;
 
+    /// This is equivalent to a `difference` between both maps.
     fn sub(self, rhs: RoaringTreemap) -> RoaringTreemap {
-        self.clone() - rhs
+        Sub::sub(self.clone(), rhs)
     }
 }
 
-impl<'a, 'b> Sub<&'a RoaringTreemap> for &'b RoaringTreemap {
+impl Sub<&RoaringTreemap> for &RoaringTreemap {
     type Output = RoaringTreemap;
 
-    fn sub(self, rhs: &'a RoaringTreemap) -> RoaringTreemap {
-        self.clone() - rhs
+    /// This is equivalent to a `difference` between both maps.
+    fn sub(self, rhs: &RoaringTreemap) -> RoaringTreemap {
+        Sub::sub(self.clone(), rhs)
     }
 }
 
 impl SubAssign<RoaringTreemap> for RoaringTreemap {
+    /// This is equivalent to a `difference` between both maps.
     fn sub_assign(&mut self, rhs: RoaringTreemap) {
-        self.difference_with(&rhs)
+        SubAssign::sub_assign(self, &rhs)
     }
 }
 
-impl<'a> SubAssign<&'a RoaringTreemap> for RoaringTreemap {
-    fn sub_assign(&mut self, rhs: &'a RoaringTreemap) {
-        self.difference_with(rhs)
+impl SubAssign<&RoaringTreemap> for RoaringTreemap {
+    /// This is equivalent to a `difference` between both maps.
+    fn sub_assign(&mut self, rhs: &RoaringTreemap) {
+        let mut keys_to_remove: Vec<u32> = Vec::new();
+        for (key, self_rb) in &mut self.map {
+            if let Some(other_rb) = rhs.map.get(key) {
+                SubAssign::sub_assign(self_rb, other_rb);
+                if self_rb.is_empty() {
+                    keys_to_remove.push(*key);
+                }
+            }
+        }
+
+        for key in keys_to_remove {
+            self.map.remove(&key);
+        }
```

Comment:
you should be able to iterate on the k/v of the rhs, and get the entry from the lhs to remove them on the fly, avoiding using a vec.

Reply:
That's indeed a lot better, thank you for that!

Reply:
BTW this same change could have been done on `BitAnd::bitand` as well but I made sure that the smallest map was iterated on, which made the algorithm faster. However, I am waiting on [the `BtreeMap::retain` method](https://doc.rust-lang.org/std/collections/btree_map/struct.BTreeMap.html#method.retain) to be available (Rust 1.53) to do this loop efficiently (loop on the smallest map) and avoid using a Vec.